# Data Analysis with Pandas:  Video Game Sales
### Paul Leonard on 11/24/2020

## Bringing in pandas and the library

In [1]:
import pandas as pd

#data from https://www.kaggle.com/gregorut/videogamesales?select=vgsales.csv
df = pd.read_csv('./vgsales.csv')

df.info()
df.head(3)
df.tail(4)
df.set_index("Name")
df[["Name","Year","Genre"]]
df.set_index("Name")[["Year","Genre"]]
df.set_index("Name")[["Year","Genre"]].sort_values('Year', ascending=True)
df['Year'] < 2000
df[df['Year'] > 2000]
df[df['Year'] > 2010].set_index("Name")[["Rank","Platform","Year"]]


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  object 
 2   Platform      16598 non-null  object 
 3   Year          16327 non-null  float64
 4   Genre         16598 non-null  object 
 5   Publisher     16540 non-null  object 
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.4+ MB


Rank Platform    Year
Name                                                                      
Grand Theft Auto V                                     17      PS3  2013.0
Grand Theft Auto V                                     24     X360  2013.0
Call of Duty: Modern Warfare 3                         30     X360  2011.0
Pokemon X/Pokemon Y                                    33      3DS  2013.0
Call of Duty: Black Ops 3                              34      PS4  2015.0
...                                                   ...      ...     ...
Rugby Challenge 3                                   16579     XOne  2016.0
Outdoors Unleashed: Africa 3D                       16581      3DS  2011.0
Fit & Fun                                           16584      Wii  2011.0
Breach                                              16588       PC  2011.0
Chou Ezaru wa Akai Hana: Koi wa Tsuki ni Shirub...  16592      PSV  2016.0

[3886 rows x 3 columns]

## What is the most common video game publisher?

In [2]:
most_common_publisher = df[["Publisher"]].mode()
most_common_publisher

Publisher
0  Electronic Arts

## What's the most common platform?

In [3]:
df[["Platform"]].mode().set_index("Platform")

Empty DataFrame
Columns: []
Index: [DS]

## What about the most common genre?

In [4]:
df[["Genre"]].mode().set_index("Genre")

Empty DataFrame
Columns: []
Index: [Action]

## What are the top 20 highest grossing games?

In [5]:
df.set_index("Name")[["Global_Sales","Year","Platform"]].sort_values('Global_Sales', ascending=False)[:20]

Global_Sales    Year Platform
Name                                                                       
Wii Sports                                           82.74  2006.0      Wii
Super Mario Bros.                                    40.24  1985.0      NES
Mario Kart Wii                                       35.82  2008.0      Wii
Wii Sports Resort                                    33.00  2009.0      Wii
Pokemon Red/Pokemon Blue                             31.37  1996.0       GB
Tetris                                               30.26  1989.0       GB
New Super Mario Bros.                                30.01  2006.0       DS
Wii Play                                             29.02  2006.0      Wii
New Super Mario Bros. Wii                            28.62  2009.0      Wii
Duck Hunt                                            28.31  1984.0      NES
Nintendogs                                           24.76  2005.0       DS
Mario Kart DS                                        23.42  2005.0       DS
Pokemon Gold/Pokemon Silver                          23.10  1999.0       GB
Wii Fit                                              22.72  2007.0      Wii
Wii Fit Plus                                         22.00  2009.0      Wii
Kinect Adventures!                                   21.82  2010.0     X360
Grand Theft Auto V                                   21.40  2013.0      PS3
Grand Theft Auto: San Andreas                        20.81  2004.0      PS2
Super Mario World                                    20.61  1990.0     SNES
Brain Age: Train Your Brain in Minutes a Day         20.22  2005.0       DS

## For North American video game sales, what’s the median?

In [6]:
df.set_index("Name")[["NA_Sales"]].median()

NA_Sales    0.08
dtype: float64

## Provide a secondary output showing ten games surrounding the median sales output

In [7]:
df[df["NA_Sales"] == 0.8].set_index("Name")[["NA_Sales","Rank"]][1:12]

NA_Sales  Rank
Name                                                              
Pinball                                                  0.8   940
Star Ocean: Till The End of Time                         0.8  1041
Call of Duty Black Ops: Declassified                     0.8  1073
Kirby & the Amazing Mirror                               0.8  1115
Tiger Woods PGA Tour 2003                                0.8  1120
Tony Hawk's American Wasteland (Old all region ...       0.8  1136
Pokemon Battle Revolution                                0.8  1141
Star Wars: The Force Unleashed II                        0.8  1205
Frogger 2: Swampy's Revenge                              0.8  1347
UFC Undisputed 2010                                      0.8  1357
Gardening Mama                                           0.8  1411

## For the top-selling game of all time, how many standard deviations above/below the mean are its sales for North America?

In [8]:
global_top_selling_all_time = df[["Global_Sales"]].max()
global_top_selling_all_time

na_std_dev = df[["NA_Sales"]].std()
na_std_dev

na_mean = df[["NA_Sales"]].mean()
na_mean

df[df["Global_Sales"] == df["Global_Sales"].max()][["Publisher","Name","Year","Genre","NA_Sales"]]
na_sales_for_top_selling_game = df[df["Global_Sales"] == df["Global_Sales"].max()][["NA_Sales"]]

std_dev_above_mean_for_na_sales = (na_sales_for_top_selling_game - na_mean) / na_std_dev
std_dev_above_mean_for_na_sales


NA_Sales
0  50.478988

## The Nintendo Wii seems to have outdone itself with games. How does its average number of sales compare with all of the other platforms?

In [9]:
overall_global_avg = df[["Global_Sales"]].mean()
wii_global_avg = df[df["Platform"] == "Wii"][["Global_Sales"]].mean()
others_global_avg = df[df["Platform"] != "Wii"][["Global_Sales"]].mean()
wii_sells_x_times_more = wii_global_avg / others_global_avg
wii_sells_x_times_more

Global_Sales    1.336297
dtype: float64

## Bonus #1:  What is the top selling game of all time?

In [10]:
df[df["Global_Sales"] == df["Global_Sales"].max()][["Publisher","Year","Genre","Name"]].set_index("Name")


Publisher    Year   Genre
Name                                
Wii Sports  Nintendo  2006.0  Sports

## Bonus #2:  List of the top ten publishers and their game counts.

In [11]:
df[["Publisher"]].value_counts()[:10]


Publisher                   
Electronic Arts                 1351
Activision                       975
Namco Bandai Games               932
Ubisoft                          921
Konami Digital Entertainment     832
THQ                              715
Nintendo                         703
Sony Computer Entertainment      683
Sega                             639
Take-Two Interactive             413
dtype: int64

## Bonus #3:  Sorted list of years with highest gloabal sales.

In [12]:
df[["Year","Global_Sales"]].groupby("Year").sum().sort_values("Global_Sales", ascending=False).head(5)

Global_Sales
Year                
2008.0        678.90
2009.0        667.30
2007.0        611.13
2010.0        600.45
2006.0        521.04

## Tests

In [13]:
def test():
    def assert_equal(actual,expected):
        assert actual ==expected, f"Expected {expected} but got {actual}"

        assert_equal(most_common_publisher, None)
        assert_equal(most_common_platform, None)
        assert_equal(most_common_genre, None)
        assert_equal(top_twenty_highest_grossing_games.iloc[0].Name, None)
        assert_equal(top_twenty_highest_grossing_games.iloc[19].Name, None)
        assert_equal(na_median_sales, None)
        assert_equal(ten_median_na_seller_names, None)

        print("Success!!!")

test()

## Extra Bonus:  What year had the highest dollar amount of global sales?  (2008 at 678.90)
### I tried to find the top sales year all in one go... but still having trouble with it.

In [14]:
df_yearly_sales = df[["Year","Global_Sales"]].groupby("Year").sum()
max_yearly_sales = df[["Year","Global_Sales"]].groupby("Year").sum().max()

# df[df_yearly_sales["Global_Sales"] == max_yearly_sales][["Year","Global_Sales"]]

df[df[["Year","Global_Sales"]].groupby("Year").sum()["Global_Sales"] == df[["Year","Global_Sales"]].groupby("Year").sum().max()["Global_Sales"]][["Year","Global_Sales"]]

# df[df[["Year","Global_Sales"]].groupby("Year").sum() == df[["Year","Global_Sales"]].groupby("Year").sum().max()]["Year"]
# df[df[["Year","Global_Sales"]].groupby("Year").sum() == df[["Year","Global_Sales"]].groupby("Year").sum().max()]
# df[df[["Year","Global_Sales"]].groupby("Year").sum() == df[["Year","Global_Sales"]].groupby("Year").sum().max()][["Year","Global_Sales"]]


IndexingError: Unalignable boolean Series provided as indexer (index of the boolean Series and of the indexed object do not match).